### PR TITLE
THOR-942 Filter Refactor (BaseNeonComponent and Subclasses)

### PR DESCRIPTION
**Pull request by ThomasSchellenbergNextCentury**
WHAT DOES THOR-942 WANT TO FIX
- Each filterable visualization component creates and maintains a list of its own filter objects (with unique and inconsistent structure) and overrides many superclass functions (getCloseableFilters, getFiltersToIgnore, getFilterText, removeFilters, setupFilters, etc.) to manage its filters.
- The visualization components depend on the neon.query library, making it impossible to use them independently of the Neon Middleware.
- Filters are no longer saved in the server (due to a previous ticket) and thus we can remove the complex callback chains from the  functions that add/delete filters.
- A lot of code to create, add, or toggle filters in the visualization components is duplicated across the codebase.
- The setupFilters function, intended to allow filters on the same field to be shared across visualizations, is clumsy, confusing, and usually implemented incorrectly (or not implemented at all).
- The visualization components show little grey boxes when filtered, and we don't want to show the boxes anymore.

HOW DOES THOR-942 FIX IT
- Created a set of interfaces in the FilterService to represent filters across the application.  We're using interfaces because they are equivalent to JSON objects and are thus compatible with future non-Typescript Web Components.  The FilterDesign interfaces are converted to Filter classes that are used internally within the FilterService.
- Replaced the many abstract filter functions in BaseNeonComponent with a single function:  designEachFilterWithNoValues.  This function allows the visualization to inform the BaseNeonComponent as to the types of filters that are compatible with it as well as a callback function to redraw its filters whenever updated.  Filters are now automatically shared across visualizations by using the output from designEachFilterWithNoValues (no need to implement setupFilters in each visualization!).
- Replaced the addFilter, removeFilter, and replaceFilter functions with deleteFilters, exchangeFilters, and toggleFilters that take FilterDesigns (JSON objects) rather that neon.query objects.
- Replaced the unique filters from the individual visualization components with a cachedFilters property in the BaseNeonComponent that copies filters from the FilterService rather than using separate (unique, inconsistent) filter objects.
- The FilterService now saves each filter not by visualization but by unique database/table/field/operator (FilterDataSource).  This way a filter can easily be shared across visualizations based on its FilterDataSource.  Filters themselves are now tied to the whole application rather than a single visualization.
- Removed the little grey boxes from the visualizations.

WHAT IS IN THIS PR
- Changes to the BaseNeonComponent itself, as detailed above.
- Updates to the filtering subclasses of BaseNeonComponent (except the Query Bar and Taxonomy Viewer, because they have their own PR) to make them work with the new FilterService and BaseNeonComponent changes.

To test this commit, please use branch THOR-942-services because that has the complete, functional set of changes.

### CODE REVIEW

File: src/app/components/annotation-viewer/annotation-viewer.component.ts [hunk 1]
```diff
@@ -200,32 +229,16 @@ export class AnnotationViewerComponent extends BaseNeonComponent implements OnIn
...
-            prettyField: part.type
-        };
-        if (this.filterIsUnique(filter) && part.annotation) {
-            this.addLocalFilter(filter);
-            let whereClause = neon.query.where(filter.field, '=', filter.value);
-            this.addNeonFilter(this.options, true, filter, whereClause);
-        }
-        if (!part.annotation) {
+        if (part.annotation) {
+            // TODO THOR-1098
```
Comment (rartiss55) [inline src/app/components/annotation-viewer/annotation-viewer.component.ts:239]: Re: THOR-1098, what did this do before/what needs to be fixed?

File: src/app/components/base-neon-component/base-neon.component.ts [hunk 2]
```diff
@@ -816,30 +766,47 @@ export abstract class BaseNeonComponent implements AfterViewInit, OnInit, OnDest
...
+            // TODO THOR-1108 Ignore filters on non-matching datastores/databases/tables.
+            this.executeAllQueryChain();
+        }
     }
 
     /**
-     * Get and configure filters from the filter service.
-     * DO NOT EXECUTE QUERIES IN THIS METHOD.
-     * This method will be called before the visualization has finished initialization!
+     * Replaces the existing global widget filters with all the compatible (matching) dash filters.
```
Comment (rartiss55) [inline src/app/components/base-neon-component/base-neon.component.ts:786]: Do we mean replacing global widget filters with what filters apply specifically for this visualization? Or am I reading this backwards?

File: src/app/components/annotation-viewer/annotation-viewer.component.ts (hunk 1 above)
Comment (ThomasSchellenbergNextCentury) [reply src/app/components/annotation-viewer/annotation-viewer.component.ts:239]: I'm not really sure what this did before.  It let you set filters on annotation "parts", whatever that means.  We're not using it now and I have no way of testing it (even if I understood what it did).  Figuring all that out seemed outside the scope of the ticket which is why I left the TODO.

File: src/app/components/base-neon-component/base-neon.component.ts (hunk 2 above)
Comment (ThomasSchellenbergNextCentury) [reply src/app/components/base-neon-component/base-neon.component.ts:786]: Yes:  Anything that's set in this widget's collection of filters will be replaced by the filters that apply specifically to this widget/visualization (a.k.a. the "compatible" filters).  Does the comment need clarification?

Comment (rartiss55) [reply src/app/components/base-neon-component/base-neon.component.ts:786]: Yeah I think some clarification here would help.

**Review by rartiss55 — CHANGES_REQUESTED** [review]
Left an earlier note to clarify something in the base-neon.component, but otherwise I think everything looks good and I tested with a couple different LORELEI datasets. The TODOs listed don't seem like high priority at the moment, and they're pretty easy to find under the 'Filter Changes' epic.

@bdpham I only checked the widgets we use for LORELEI, so wanted to leave a note for you to be sure to test out the changes made in the network graph, thumbnail grid, newsfeed and timeline.

**Pull request merged**